signed-note.md

```diff
@@ -0,0 +1,138 @@
+(U+002B) characters.
+
+Signatures are the base64 encoding of 4+n bytes. The first four bytes in the
+signature are the uint32 key ID stored in big-endian order. The remaining n
+bytes are the result of using the specified key to sign the note text, according
+to the signature algorithm.
+
+Verifiers SHOULD apply a maximum limit to the number of signatures in a note
+(or, equivalently, to the overall size of a note), but MUST accept at least up
+to 16 signatures. Remember that post-quantum signatures can be up to nearly 5kB
+in size.
+
+## Signatures
+
+    — <key name> base64(32-bit key ID || signature)
+
+The tuple of key name and key ID uniquely identifies the key that performed the
+signature and the algorithm it used. Verifiers MUST ignore signatures from
+unknown keys, even if they share a name or ID (but not both) with a known key.
+If the key name and ID match a known key, verifiers MUST verify the signature.
+If a signature from a known key fails to verify, clients MAY reject the whole
+note. If no signature from a known key verifies successfully, clients MUST
+reject the note.
+
+Note that the key ID is intentionally short, as it is an identifier, and not a
+cryptographically strong hash. An attacker can easily compute a public key that
+will collide a target key hash, but all they will get is a signature
+verification failure, because the client was necessarily securely configured
+with the expected public key for the key hash.
+
+It is RECOMMENDED that key names be schema-less URLs (e.g. `example.com/abc123`),
+to avoid accidental collisions. Those endpoints don’t need to be reachable over
+the network.
+
+It is RECOMMENDED that key IDs be the first four bytes (interpreted in
+big-endian order) of the SHA-256 hash of the following sequence: the key name, a
+newline, a single signature type identifier byte assigned below, and the
+encoding of the public key. Note that this allows reusing the same cryptographic
+key and name for different signature types without ambiguity, if necessary (and
+if domain separation can be ensured).
+
+    key ID = SHA-256(key name || 0x0A || signature type || public key)[:4]
+
+Accidental collisions of *both* key name and key ID can lead to errors if
+signatures from colliding keys pairs are applied to the same note. The
+recommendations above make such collisions exceedingly unlikely.
+
+### Signature types
+
+The following are the assigned signature type identifier bytes. Other
+specifications may define the format and semantics of these signature types, and
+they MAY require specific content structures of the note text.
+
+Although multiple signature types are specified, implementations SHOULD select
+only the one(s) required for their design, and avoid supporting multiple ones at
+runtime if possible.
+
+* `0x01` — Ed25519 signatures as specified by this document.
+
+* `0x02` — ECDSA signatures as implemented by github.com/transparency-dev/witness.
+
```

Comment:
Could RSA be supported? While Sigstore defaults to ECDSA, this would be helpful for private deployments of Sigstore infrastructure.

Reply:
We can specify an RSA note signature format (although we'll want to pick specific parameters, like RSA-PSS, hash, key size requirements, etc.) but every new log signature algorithm needs to be supported by every witness, and like everything with cryptography, adding more complexity adds more things that can go wrong.

What makes private deployments unable to use ECDSA? It's supported by every HSM, KMS, and compliance regime I am aware of by now.

Reply:
(Merging this PR now, so feel free to pull this out into an issue!)

Reply:
Opened https://github.com/C2SP/C2SP/issues/63 to discuss more, thanks!